pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1721,10 +1705,12 @@ private void closeProducerTasks() {
             sendTimeout = null;
         }
 
-        ScheduledFuture<?> batchTimerTask = this.batchTimerTask;
-        if (batchTimerTask != null) {
-            batchTimerTask.cancel(false);
-            this.batchTimerTask = null;
+        synchronized (this) {
+            ScheduledFuture<?> batchTimerTask = this.batchTimerTask;
+            if (batchTimerTask != null) {
+                batchTimerTask.cancel(false);
+                this.batchTimerTask = null;
+            }
         }
```

Comment:
Since this is no more a "timer task", it would be better to let the task complete gracefully. It needs some timeout when waiting for the completion.

Reply:
Are you suggesting we remove this from the close block? I'm fine with that since it will already exit early based on the producer's state.

Reply:
Yes, removing it from the close block or waiting for the current task to complete. Why would you cancel it? What is the desired behavior in cancelling?